appfl/algorithm/fedavg.py

```diff
@@ -4,71 +4,97 @@
 
 from collections import OrderedDict
 from .algorithm import BaseServer, BaseClient
+from .misc import *
 
 import torch
+from torch.optim import *
 from torch.nn import CrossEntropyLoss
 from torch.utils.data import DataLoader
+import copy
+
 
 class FedAvgServer(BaseServer):
-    def __init__(self, model, num_clients, device, **kwargs):
-        super(FedAvgServer, self).__init__(model, num_clients, device)
+    def __init__(self, weights, model, num_clients, device, **kwargs):
+        super(FedAvgServer, self).__init__(weights, model, num_clients, device)
 
         self.__dict__.update(kwargs)
+        self.num_clients = num_clients
+        self.weights = weights
+
+        self.primal_states = OrderedDict()
+        for i in range(num_clients):
+            self.primal_states[i] = OrderedDict()
+
+    def update(self, local_states: OrderedDict):
+
+        """ Inputs """
+        global_state = OrderedDict()
+        primal_recover_from_local_states(self, local_states)
+        
 
-    # update global model
-    def update(self, global_state: OrderedDict, local_states: OrderedDict):
-        update_state = OrderedDict()
-        for k, state in local_states.items():
-            for key in self.model.state_dict().keys():
-                if key in update_state.keys():
-                    update_state[key] += state[key] / self.num_clients
-                else:
-                    update_state[key] = state[key] / self.num_clients
+        """ Outputs """
```

Comment:
Inputs and Outputs..? Can you give us some more descriptions?

Reply:
Fixed